analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/LoopDetector.cs

```diff
@@ -83,11 +83,22 @@ void MergeWithIntersectingLoops(int[] path, int last)
         }
     }
 
-    private static IEnumerable<int> Successors(ControlFlowGraph cfg, int block)
+    private static IEnumerable<int> Successors(ControlFlowGraph cfg, int index)
     {
-        var successors = cfg.Blocks[block].SuccessorBlocks.Select(x => x.Ordinal);
-        return cfg.Blocks[block].EnclosingRegion(ControlFlowRegionKind.Try) is { } tryRegion
-            ? successors.Concat(tryRegion.ReachableHandlers().Select(x => x.FirstBlockOrdinal))
-            : successors;
+        var block = cfg.Blocks[index];
+        var successors = block.SuccessorBlocks.Select(x => x.Ordinal);
+        if (block.EnclosingRegion(ControlFlowRegionKind.Try) is { } tryRegion)
+        {
+            successors = successors.Concat(tryRegion.ReachableHandlers().Select(x => x.FirstBlockOrdinal));
+        }
+        if (block.EnclosingRegion(ControlFlowRegionKind.Finally) is { } finallyRegion)
```

Comment:
Imagine this scenario, while asking for successors of `Where does this go?`

We miss-match `finallyRegion` here, and compute wrong results. This should use the infamous `EnclosingNonLocalLifetimeRegion` to find it's nearest `Finally` region.

Let's try to craft UT what would break this... like nest this in a loop that will just `break;` on most of the paths.
```
Try
  Console.WriteLine("Ignore me")
Finally
   Try
      Console.WriteLine("Where does this go?")
   Finally
   End Try
   Console.WriteLine("It should go only here")
End Try
Console.WriteLine("But it will end up here as well")
```

Reply:
I don't think it matters in terms of loop detection. The code will reach the last block after the finally was fully executed (just not directly). For our use case, it is only relevant whether or not the block is reachable. It doesn't matter if the connection is direct or not. Thus, having both paths doesn't hurt.

Reply:
Once you start adding GoTo to form a loop, it can start making wrong results. That's why I wanted to try to craft UT from that

Reply:
Due to `error CS0157: Control cannot leave the body of a finally clause` (`Control` being `goto`), I cannot come up with a failing example.
Regardless: How would you propose to implement it differently?